File: lib/models/counter.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:which/models/question.dart';

@immutable
class Counter {
  static const int shardsNum = 10;
  final String questionId;
  final int readed;
  final int watched;
  final int answer1;
  final int answer2;

  const Counter({
    required this.questionId,
    required this.readed,
    required this.watched,
    required this.answer1,
    required this.answer2,
  });

  static Map<String, dynamic> forIncrement(
    Question question, {
    bool incrementRead = false,
    bool incrementWatch = false,
    bool incrementAnswer1 = false,
    bool incrementAnswer2 = false,
  }) {
    final Map<String, dynamic> data = {'questionId': question.questionId};
    if (incrementRead) data['readed'] = FieldValue.increment(1);
    if (incrementWatch) data['watched'] = FieldValue.increment(1);
    if (incrementAnswer1) data['answer1'] = FieldValue.increment(1);
    if (incrementAnswer2) data['answer2'] = FieldValue.increment(1);
    return data;
  }

  factory Counter.fromDocs(List<DocumentSnapshot> docs) {
    String questionId = '';
    int readed = 0;
    int watched = 0;
    int answer1 = 0;
    int answer2 = 0;
    for (final DocumentSnapshot doc in docs) {
      final Map<String, dynamic> data = doc.data() as Map<String, dynamic>;
      questionId = data['questionId'] as String? ?? '';
      readed += data['readed'] as int? ?? 0;
      watched += data['watched'] as int? ?? 0;
      answer1 += data['answer1'] as int? ?? 0;
      answer2 += data['answer2'] as int? ?? 0;
    }
    return Counter(
      questionId: questionId,
      readed: readed,
      watched: watched,
      answer1: answer1,
      answer2: answer2,
    );
  }

  // copyWith
  Counter copyWith({
    String? questionId,
    int? readed,
    int? watched,
    int? answer1,
    int? answer2,
  }) {
    return Counter(
      questionId: questionId ?? this.questionId,
      readed: readed ?? this.readed,
      watched: watched ?? this.watched,
      answer1: answer1 ?? this.answer1,
      answer2: answer2 ?? this.answer2,
    );
  }

  // hashCode
  @override
  int get hashCode =>
      questionId.hashCode ^
      readed.hashCode ^
      watched.hashCode ^
      answer1.hashCode ^
      answer2.hashCode;

  // ==
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Counter &&
          runtimeType == other.runtimeType &&
          questionId == other.questionId &&
          readed == other.readed &&
          watched == other.watched &&
          answer1 == other.answer1 &&
          answer2 == other.answer2;
}

File: functions/src/models/counter.ts
/**
 * Counterクラスは質問に対する統計情報を保持します。
 */
export class Counter {
  questionId: string;
  readed: number;
  watched: number;
  answer1: number;
  answer2: number;

  /**
   * Counterのコンストラクタ
   * @param {string} questionId - 質問のID
   * @param {number} readed - 読まれた回数
   * @param {number} watched - 見られた回数
   * @param {number} answer1 - 回答1の数
   * @param {number} answer2 - 回答2の数
   */
  constructor(
    questionId: string,
    readed: number,
    watched: number,
    answer1: number,
    answer2: number
  ) {
    this.questionId = questionId ?? "";
    this.readed = readed ?? 0;
    this.watched = watched ?? 0;
    this.answer1 = answer1 ?? 0;
    this.answer2 = answer2 ?? 0;
  }

  /**
   * ドキュメントの配列からCounterインスタンスを生成します。
   * @param {string} questionId - 質問のID
   * @param {Array<FirebaseFirestore.QueryDocumentSnapshot>} docs - ドキュメントの配列
   * @return {Counter} Counterインスタンス
   */
  static fromDocs(
    questionId: string,
    docs: Array<FirebaseFirestore.QueryDocumentSnapshot>
  ): Counter {
    let readed = 0;
    let watched = 0;
    let answer1 = 0;
    let answer2 = 0;

    docs.forEach((doc) => {
      const data = doc.data();
      if (data.questionId === questionId) {
        readed += data.readed ?? 0;
        watched += data.watched ?? 0;
        answer1 += data.answer1 ?? 0;
        answer2 += data.answer2 ?? 0;
      }
    });

    return new Counter(questionId, readed, watched, answer1, answer2);
  }

  /**
   * 人気度を取得します。
   * @return {number} 人気度
   */
  getPopular(): number {
    const watched = this.watched;
    const readed = this.readed;
    if (watched >= readed) return 1;
    return watched / readed;
  }
}
